control-plane/connect-inject/terminatinggatewayservice_controller.go

```diff
@@ -0,0 +1,515 @@
+
+	// The DeletionTimestamp is zero when the object has not been marked for deletion. The finalizer is added
+	// in case it does not exist to all resources. If the DeletionTimestamp is non-zero, the object has been
+	// marked for deletion and goes into the deletion workflow.
+	if terminatingGatewayService.GetDeletionTimestamp().IsZero() {
+		if !controllerutil.ContainsFinalizer(terminatingGatewayService, FinalizerName) {
+			controllerutil.AddFinalizer(terminatingGatewayService, FinalizerName)
+			if err := r.Update(ctx, terminatingGatewayService); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+	} else {
+		if containsString(terminatingGatewayService.Finalizers, FinalizerName) {
+			r.Log.Info("TerminatingGatewayService was deleted, deleting from consul", "name", req.Name, "ns", req.Namespace)
+			_, err := r.deleteService(spec.Service.ServiceName)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+			controllerutil.RemoveFinalizer(terminatingGatewayService, FinalizerName)
+			err = r.Update(ctx, terminatingGatewayService)
+			return ctrl.Result{}, err
+		}
+	}
+
+	err = r.createOrUpdateService(terminatingGatewayService, ctx)
+	if err != nil {
+		r.Log.Error(err, "Unable to create or update service", "name", req.Name, "ns", req.Namespace)
+
+		err = r.updateStatusError(ctx, terminatingGatewayService, err)
+		if err != nil {
+			r.Log.Error(err, "failed to update TerminatingGatewayService status", "name", terminatingGatewayService.Name, terminatingGatewayService.Namespace)
+		}
+	}
+
+	return ctrl.Result{}, err
+}
+
+// terminatingGatewayACLRole returns the ACL role of the running terminating gateway.
+func terminatingGatewayACLRole(aclRoleList []*api.ACLRole) (*api.ACLRole, error) {
+	strToFind := "terminating-gateway"
+
+	result := &api.ACLRole{}
+	roleFound := false
+
+	for _, role := range aclRoleList {
+		if strings.Contains(role.Name, strToFind) {
+			result = role
+			roleFound = true
+			break
+		}
+	}
+
+	if !roleFound {
+		return result, errors.New("terminating Gateway ACL Role not found")
+	}
+	return result, nil
+}
+
+// updateStatus updates the terminatingGatewayService's information in the status.
+func (r *TerminatingGatewayServiceController) updateStatus(ctx context.Context, terminatingGatewayService *consulv1alpha1.TerminatingGatewayService) error {
+
+	policyName := ""
+	if r.AclEnabled {
+		policyName = fmt.Sprintf("%s-write-policy", terminatingGatewayService.Spec.Service.ServiceName)
+	}
+
+	terminatingGatewayService.Status.LastSyncedTime = &metav1.Time{Time: time.Now()}
+	terminatingGatewayService.SetSyncedCondition(corev1.ConditionTrue, "", "")
+
+	terminatingGatewayService.Status.ServiceInfoRef = &consulv1alpha1.ServiceInfoRefStatus{
+		ServiceName: terminatingGatewayService.Spec.Service.ServiceName,
+		PolicyName:  policyName,
+	}
+
+	err := r.Status().Update(ctx, terminatingGatewayService)
+	if err != nil {
+		err = fmt.Errorf("failed to update TerminatingGatewayService status: %v", err)
```

Comment:
When you get the err, you should probably return right away. The same is true in updateStatusError(). So it should be:

```
	err := r.Status().Update(ctx, terminatingGatewayService)
	if err != nil {
		return fmt.Errorf("failed to update TerminatingGatewayService status: %v", err)
	}
	return nil
```

Returning right away is the more `idiomatic` way of writing go code and it can prevent subtle bugs.

Reply:
Done